### PR TITLE
no more typeid, it's constexpr

### PR DESCRIPTION
**Pull request by al-tu**
isSamePointType function 
1. is constexpr now
2. uses type traits instead of typeid comparison
correspondence estimation functions forward point selection to helper template functions specializations instead of if blocks with runtime type checks

### CODE REVIEW

File: common/include/pcl/common/io.h [hunk 1]
```diff
@@ -319,10 +319,10 @@ namespace pcl
                   pcl::PCLPointCloud2 &cloud_out);
 
   /** \brief Check if two given point types are the same or not. */
-  template <typename Point1T, typename Point2T> inline bool
+  template <typename Point1T, typename Point2T> inline constexpr bool
```
Comment (al-tu) [inline common/include/pcl/common/io.h:322]: should this function live in `common/include/pcl/type_traits.h` perhaps?

File: registration/include/pcl/registration/impl/correspondence_estimation.hpp [hunk 2]
```diff
@@ -120,50 +149,31 @@ CorrespondenceEstimation<PointSource, PointTarget, Scalar>::determineCorresponde
   if (!initCompute())
     return;
 
-  double max_dist_sqr = max_distance * max_distance;
-
   correspondences.resize(indices_->size());
 
+  pcl::Correspondence corr;
```
Comment (kunaltyagi) [inline registration/include/pcl/registration/impl/correspondence_estimation.hpp:154]: Why are there some unrelated code modifications?

File: registration/include/pcl/registration/impl/correspondence_estimation.hpp [hunk 3]
```diff
@@ -120,50 +149,31 @@ CorrespondenceEstimation<PointSource, PointTarget, Scalar>::determineCorresponde
...
-      corr.distance = distance[0];
-      correspondences[nr_valid_correspondences++] = corr;
-    }
+  // Iterate over the input set of source indices
+  for (const auto& idx : (*indices_)) {
+    // Check if the template types are the same. If true, avoid a copy.
+    // Both point types MUST be registered using the POINT_CLOUD_REGISTER_POINT_STRUCT
+    // macro!
+    const auto& pt(
+        detail::pointCopyOrRef<PointSource, PointTarget, decltype(idx)>(input_, idx));
```
Comment (kunaltyagi) [inline registration/include/pcl/registration/impl/correspondence_estimation.hpp:166]: Would this be possible?
```suggestion
    const auto& pt =
        detail::pointCopyOrRef<PointSource, PointTarget, decltype(idx)>(input_, idx);
```

Comment (kunaltyagi) [reply registration/include/pcl/registration/impl/correspondence_estimation.hpp:166]: And in other places too

File: registration/include/pcl/registration/impl/correspondence_estimation.hpp [hunk 4]
```diff
@@ -112,6 +112,35 @@ CorrespondenceEstimationBase<PointSource, PointTarget, Scalar>::initComputeRecip
   return (true);
 }
 
+namespace detail {
+
+template <
+    typename PointSource,
+    typename PointTarget,
+    typename Index,
```
Comment (kunaltyagi) [inline registration/include/pcl/registration/impl/correspondence_estimation.hpp:120]: If we put PointTarget as 1st argument, the rest can be auto-deduced based on function argument.

File: test/common/test_common.cpp [hunk 5]
```diff
@@ -360,27 +360,27 @@ TEST (PCL, SetIfFieldExists)
...
-  EXPECT_FALSE (status);
-  status = isSamePointType<PointNormal, PointNormal> ();
-  EXPECT_TRUE (status);
-  status = isSamePointType<PointNormal, PointXYZRGBNormal> ();
-  EXPECT_FALSE (status);
-  status = isSamePointType<PointXYZRGB, PointXYZRGB> ();
-  EXPECT_TRUE (status);
-
+  constexpr const bool status0 = isSamePointType<PointXYZ, PointXYZ> ();
+  EXPECT_TRUE (status0);
```
Comment (kunaltyagi) [inline test/common/test_common.cpp:367]: Thoughts on combining into a single line?

File: common/include/pcl/common/io.h (hunk 1 above)
Comment (kunaltyagi) [reply common/include/pcl/common/io.h:322]: We can simplify this a lot more (and deprecate it) since this is just a function call instead of `std::is_same_v<A, B>`. Thoughts: @mvieth ?

Comment (mvieth) [reply common/include/pcl/common/io.h:322]: I am not necessarily against deprecating `isSamePointType`, but I don't really see a problem with people using `pcl::isSamePointType<A, B>()` instead of `std::is_same<A, B>::value` (or `is_same_v` starting with C++17).
How would you simplify it?

Comment (kunaltyagi) [reply common/include/pcl/common/io.h:322]: Alright. If we don't deprecate it, should we move it to `type_traits.h` or some other file than `io.h`?

File: registration/include/pcl/registration/impl/correspondence_estimation.hpp (hunk 2 above)
Comment (al-tu) [reply registration/include/pcl/registration/impl/correspondence_estimation.hpp]: fixed

File: registration/include/pcl/registration/impl/correspondence_estimation.hpp (hunk 4 above)
Comment (al-tu) [reply registration/include/pcl/registration/impl/correspondence_estimation.hpp]: done

File: registration/include/pcl/registration/impl/correspondence_estimation.hpp (hunk 3 above)
Comment (al-tu) [reply registration/include/pcl/registration/impl/correspondence_estimation.hpp]: sure, done

File: test/common/test_common.cpp (hunk 5 above)
Comment (al-tu) [reply test/common/test_common.cpp]: done
also, only two of these tests seem like relevant, should i remove duplication, eg there's no point to test
`EXPECT_TRUE (isSamePointType<PointNormal, PointNormal> ());`
if there's
`EXPECT_TRUE (isSamePointType<PointXYZ, PointXYZ> ());`
what do you think?

File: registration/include/pcl/registration/impl/correspondence_estimation_backprojection.hpp [hunk 6]
```diff
@@ -188,98 +144,49 @@ CorrespondenceEstimationBackProjection<PointSource, PointTarget, NormalT, Scalar
...
-        if (dist < min_dist) {
-          min_dist = dist;
-          min_index = static_cast<int>(j);
-        }
+  // Iterate over the input set of source indices
+  for (const auto& idx_i : (*indices_)) {
+    // Check if the template types are the same. If true, avoid a copy.
+    // Both point types MUST be registered using the POINT_CLOUD_REGISTER_POINT_STRUCT
+    // macro!
+    const auto& pt = detail::pointCopyOrRef<PointSource, PointTarget, decltype(idx_i)>(
```
Comment (kunaltyagi) [inline registration/include/pcl/registration/impl/correspondence_estimation_backprojection.hpp:152]: Do we need the long form here?

File: registration/include/pcl/registration/impl/correspondence_estimation.hpp [hunk 7]
```diff
@@ -192,59 +201,31 @@ CorrespondenceEstimation<PointSource, PointTarget, Scalar>::
...
+
+    const auto& pt_src = detail::pointCopyOrRef<PointTarget>(input_, idx);
+
+    tree_->nearestKSearch(pt_src, 1, index, distance);
+    if (distance[0] > max_dist_sqr)
+      continue;
+
+    target_idx = index[0];
+    const auto& pt_tgt =
+        detail::pointCopyOrRef<PointTarget, PointSource, decltype(target_idx)>(
```
Comment (kunaltyagi) [inline registration/include/pcl/registration/impl/correspondence_estimation.hpp:218]: Same here and in other places

File: registration/include/pcl/registration/impl/correspondence_estimation_backprojection.hpp (hunk 6 above)
Comment (al-tu) [reply registration/include/pcl/registration/impl/correspondence_estimation_backprojection.hpp]: revisited

Comment (al-tu) [reply registration/include/pcl/registration/impl/correspondence_estimation_backprojection.hpp]: while it is possible to deduce `Index` type, it's not true for `PointSource` (second type parameter): type cannot be infered as it is used in nested-type-specifier

File: common/include/pcl/common/io.h (hunk 1 above)
Comment (kunaltyagi) [reply common/include/pcl/common/io.h:322]: Apart from this, LGTM

**Review by kunaltyagi — APPROVED** [review]
If CI is 🟢

File: test/common/test_common.cpp (hunk 5 above)
Comment (al-tu) [reply test/common/test_common.cpp]: hi! should i revert my changes in tests?

Comment (mvieth) [reply test/common/test_common.cpp]: Sorry for the delay.

> someone reading those tests would (mistakenly) assume that there is some difference in comparison logic for PointXYZ and PointNormal types

I don't really agree with that thought. I think it is normal to test a function with many different parameters, also template parameters, without any implication that the function works differently for some parameters. I think the `EXPECT_TRUE` and `EXPECT_FALSE` also make clear what the function should do in each case. If you want to add a clarifying comment somewhere, that is of course always a good idea. I just think removing already existing tests is counter-intuitive, considering that too much code in PCL is untested.

**Comment by mvieth** [comment]
@al-tu The new test(s) don't seem to compile yet

File: registration/include/pcl/registration/impl/correspondence_estimation_normal_shooting.hpp [hunk 8]
```diff
@@ -199,110 +156,58 @@ CorrespondenceEstimationNormalShooting<PointSource, PointTarget, NormalT, Scalar
...
-      corr.distance = nn_dists[min_index]; // min_dist;
-      correspondences[nr_valid_correspondences++] = corr;
     }
+    if (min_dist > max_distance)
+      continue;
+
+    // Check if the correspondence is reciprocal
+    target_idx = nn_indices[min_index];
+    tree_reciprocal_->nearestKSearch(
+        (*target_)[target_idx], 1, index_reciprocal, distance_reciprocal);
```
[the remark this answers is not in the document]
Comment (al-tu) [reply registration/include/pcl/registration/impl/correspondence_estimation_normal_shooting.hpp]: suppose that you're right, it seems like only the cases when point types are the same were covered, i've added another one for completeness (though ), so it should be correct now, but that required some additional changes
PTAL, i might have messed up that source-target logic (hope not)

**Comment by al-tu** [comment]
@kunaltyagi @mvieth hello! do you have any idea why [one particular test](https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=21038&view=results) on clang/macos catalina fails? may this be a flaky situation? it would be nice to be able to restart ci via bot command

**Comment by mvieth** [comment]
> @kunaltyagi @mvieth hello! do you have any idea why [one particular test](https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=21038&view=results) on clang/macos catalina fails? may this be a flaky situation? it would be nice to be able to restart ci via bot command

I restarted the CI. Unfortunately, there are a few tests that fail randomly (but rarely)

**Comment by al-tu** [comment]
> > @kunaltyagi @mvieth hello! do you have any idea why [one particular test](https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=21038&view=results) on clang/macos catalina fails? may this be a flaky situation? it would be nice to be able to restart ci via bot command
> 
> I restarted the CI. Unfortunately, there are a few tests that fail randomly (but rarely)

thanks a lot! passes now, so yes, that was a random fail

**Comment by al-tu** [comment]
hello. it's actually ready now, could you please re-review? thanks!

**Comment by larshg** [comment]
@al-tu Do you have time to rebase this?

File: test/registration/test_correspondence_estimation.cpp [hunk 9]
```diff
@@ -41,34 +41,88 @@
...
   // Defining two parallel planes differing only by the y co-ordinate
-  for (float i = 0.0f; i < 10.0f; i += 0.2f)
+  for (std::size_t i = 0; i < 50; ++i)
   {
-    for (float z = 0.0f; z < 5.0f; z += 0.2f)
+    for (std::size_t j = 0; j < 25; ++j)
     {
-      cloud1->points.emplace_back(i, 0, z);
-      cloud2->points.emplace_back(i, 2, z); // Ideally this should be the corresponding point to the point defined in the previous line
+      cloud1->points.push_back(makePointWithParams<PointSource>(i * 0.2f, 0.f, j * 0.2f));
```
Comment (larshg) [inline test/registration/test_correspondence_estimation.cpp:109]:
```suggestion
      cloud1->push_back(makePointWithParams<PointSource>(i * 0.2f, 0.f, j * 0.2f));
```

File: test/registration/test_correspondence_estimation.cpp [hunk 10]
```diff
@@ -41,34 +41,88 @@
...
+      cloud2->points.push_back(makePointWithParams<PointTarget>(i * 0.2f, 2.f, j * 0.2f)); // Ideally this should be the corresponding point to the point defined in the previous line
```
Comment (larshg) [inline test/registration/test_correspondence_estimation.cpp:110]:
```suggestion
      cloud2->push_back(makePointWithParams<PointTarget>(i * 0.2f, 2.f, j * 0.2f)); // Ideally this should be the corresponding point to the point defined in the previous line
```

File: test/common/test_io.cpp [hunk 11]
```diff
@@ -51,29 +51,6 @@ PointXYZRGBA pt_xyz_rgba, pt_xyz_rgba2;
 PointXYZRGB pt_xyz_rgb;
 PointXYZ pt_xyz;
 
-///////////////////////////////////////////////////////////////////////////////////////////
```
Comment (larshg) [inline test/common/test_io.cpp:54]: Should we keep these tests? @mvieth

**Comment by al-tu** [comment]
> @al-tu Do you have time to rebase this?

sure, done

File: test/registration/test_correspondence_estimation.cpp (hunk 10 above)
Comment (al-tu) [reply test/registration/test_correspondence_estimation.cpp]: fixed

File: test/registration/test_correspondence_estimation.cpp (hunk 9 above)
Comment (al-tu) [reply test/registration/test_correspondence_estimation.cpp]: fixed

File: test/common/test_io.cpp (hunk 11 above)
Comment (mvieth) [reply test/common/test_io.cpp:54]: These tests are duplicated from test/common/test_common.cpp (https://github.com/PointCloudLibrary/pcl/blob/master/test/common/test_common.cpp#L364)

Comment (larshg) [reply test/common/test_io.cpp:54]: Ahh, alright.

**Review by mvieth — APPROVED** [review]
Thank you, and sorry that we took so long with the reviews

**Review by larshg — APPROVED** [review]
